finat/bell.py

```diff
@@ -26,63 +26,63 @@ def basis_transformation(self, coordinate_mapping):
         for multiindex in numpy.ndindex(V.shape):
             V[multiindex] = Literal(V[multiindex])
 
+        voffset = 6
         for v in range(3):
-            s = 6*v
+            s = voffset * v
             V[s, s] = Literal(1)
             for i in range(2):
                 for j in range(2):
-                    V[s+1+i, s+1+j] = J[j, i]
+                    V[s+1+j, s+1+i] = J[j, i]
             V[s+3, s+3] = J[0, 0]*J[0, 0]
-            V[s+3, s+4] = 2*J[0, 0]*J[1, 0]
-            V[s+3, s+5] = J[1, 0]*J[1, 0]
-            V[s+4, s+3] = J[0, 0]*J[0, 1]
+            V[s+4, s+3] = 2*J[0, 0]*J[1, 0]
+            V[s+5, s+3] = J[1, 0]*J[1, 0]
+            V[s+3, s+4] = J[0, 0]*J[0, 1]
             V[s+4, s+4] = J[0, 0]*J[1, 1] + J[1, 0]*J[0, 1]
-            V[s+4, s+5] = J[1, 0]*J[1, 1]
-            V[s+5, s+3] = J[0, 1]*J[0, 1]
-            V[s+5, s+4] = 2*J[0, 1]*J[1, 1]
+            V[s+5, s+4] = J[1, 0]*J[1, 1]
+            V[s+3, s+5] = J[0, 1]*J[0, 1]
+            V[s+4, s+5] = 2*J[0, 1]*J[1, 1]
             V[s+5, s+5] = J[1, 1]*J[1, 1]
 
         for e in range(3):
-            v0id, v1id = [i for i in range(3) if i != e]
+            s = 3 * voffset + e
+            v0id, v1id = [i * voffset for i in range(3) if i != e]
 
             # nhat . J^{-T} . t
             foo = (rns[e, 0]*(J[0, 0]*pts[e, 0] + J[1, 0]*pts[e, 1])
                    + rns[e, 1]*(J[0, 1]*pts[e, 0] + J[1, 1]*pts[e, 1]))
 
             # vertex points
-            V[18+e, 6*v0id] = -1/21 * (foo / pel[e])
-            V[18+e, 6*v1id] = 1/21 * (foo / pel[e])
+            V[v0id, s] = -1/21 * (foo / pel[e])
+            V[v1id, s] = 1/21 * (foo / pel[e])
 
             # vertex derivatives
-            for i in (0, 1):
-                V[18+e, 6*v0id+1+i] = -1/42*foo*pts[e, i]
-                V[18+e, 6*v1id+1+i] = V[18+e, 6*v0id+1+i]
+            for i in range(2):
+                V[v0id+1+i, s] = -1/42*foo*pts[e, i]
+                V[v1id+1+i, s] = V[v0id+1+i, s]
 
             # second derivatives
             tau = [pts[e, 0]*pts[e, 0],
                    2*pts[e, 0]*pts[e, 1],
                    pts[e, 1]*pts[e, 1]]
 
-            for i in (0, 1, 2):
-                V[18+e, 6*v0id+3+i] = -1/252 * (pel[e]*foo*tau[i])
-                V[18+e, 6*v1id+3+i] = 1/252 * (pel[e]*foo*tau[i])
+            for i in range(3):
+                V[v0id+3+i, s] = -1/252 * (pel[e]*foo*tau[i])
+                V[v1id+3+i, s] = 1/252 * (pel[e]*foo*tau[i])
 
         h = coordinate_mapping.cell_size()
 
         for v in range(3):
+            s = voffset * v
             for k in range(2):
-                for i in range(21):
-                    V[i, 6*v+1+k] = V[i, 6*v+1+k] / h[v]
+                V[s+1+k, :] /= h[v]
             for k in range(3):
-                for i in range(21):
-                    V[i, 6*v+3+k] = V[i, 6*v+3+k] / (h[v]*h[v])
+                V[s+3+k, :] /= (h[v]*h[v])
 
-        return ListTensor(V.T)
+        return ListTensor(V)
```

Comment:
This is unrelated to macroelements and just constructing `V.T` directly?

Reply:
Correct, I took Bell as template for HCT but I also cleaned up Bell

Reply:
Bell is not the right go-by.  It was implemented based on Argyris, which used the point normal for edge dof.  It's actually just on a branch (need to fix this!) but Wu-Xu used moments of normal derivatives, which turn out much cleaner.  The Wu-Xu element has all the dofs we need and some extra ones (second normal derivatives on edges), so this is the right go-by for us (see recent paper by Bock/Cotter/Kirby in CAMWA for a write-up).  However, that plus some aspects from Bell would get us reduced HCT.